<a href="https://colab.research.google.com/github/LarrySerratos/CienciaMatematicadeDatos/blob/main/Examen_Diplomado_Ciencia_de_Datos_Larry_Serratos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DIPLOMADO CIENCIA MATEMÁTICA DE DATOS

## Examen

### Examen Teórico

**Instrucciones**

Contesta las siguientes.

1. ¿Cuál es la diferencia entre Data Analysis y Data Science?

**R.** Data Analysis se concentra en la descripcion y visualizacion de los datos para encontrar insights o respuestas a preguntas de utilidad para tomar desiciones. Data Science genera modelos y algoritsmos para predecir y automatizar procesos.

2. ¿Cuál es el proposito del análisis exploratorio?

**R.** En general, conocer la estructura y caracteristicas de un conjunto de datos. Conocer las variables y tipos de datos, comprender su distribucion, identificar patrones y valores atipicos. Para evaluar la calidad de los datos, formular hipotesis y preparar los datos y tomar desiciones para analisis o modelos posteriores.

3. ¿Cuáles son los problemas más comunes que uno encuentra en el análisis de datos?

 **R.**  La falta de datos, los outliers, datos duplicados, datos redundantes, la falta de formato, falta de claridad en la descripcion, sesgos, y sobrecarga de datos.

4. ¿Cuáles son los pasos en un proyecto de análisis de datos?


**R.**  
- Definición del problema o preguntas a resolver
- Limpieza de datos (en caso que no exista la base de datos aun)
- Preparación de los datos
- Análisis exploratorio (EDA)
- Modelado de Datos
- Interpretación
- Comunicación de resultados

5. ¿Cuáles son las diferencias en Aprendizaje Supervisado y Aprendizaje No Supervisado?

**R.**  El modelo supervisado se entrena con datos etiquetados, es decir, conociendo las respuestas correctas, se usa para modelos predictivos y de clasificación. El modelo no supervisado es entrenado sin etiquetas, solo se conocen los valores de entrada, se usa para descubrir patrones y reducir dimensionalidades.

6. ¿Qué es Reducción de Dimensiones y cuáles son sus beneficios?

**R.** Se trata de reducir las variables de un conjunto de datos transformandolos a un espacio de menor dimension, conservando la mayor parte de la información relevante. Entre los beneficios son la eficiencia computacional, al reducir el volumen se aceleran los procesos de entrenamiento y evaluacion de modelos, se eliminan variables irrelevantes o redundantes, se previene el sobre ajuste y se facilita la visualizacion e interpretabilidad.

7. ¿Qué son los sistemas de recomendación?

**R.** Es un modelo que analiza patrones apartir de los datos de un usuario y ususarios similares similares para prededir sus las preferencias y hacer recomendaciones personalizadas que sean atractivas para este usuario. Por ejemplo recomendacion de peliculas, musica, videos, productos.

8. ¿Qué es clustering y para qué sirve en la ciencia de datos?

**R.** Es una tecnica no supervisada que agrupa datos de manera que los datos en cada grupo son mas similares entre ellos que con los datos de otros grupos. Se utiliza para descubrir anomalias, estructuras, reduccion de dimensionalidad, etc.

9. ¿Cómo puedo elegir el mejor valor de K para K-Means?

**R.** En el curso vimos el metodo del codo. Se calcula la suma de las distancias cuadradas dentro del cluster, y se grafican respecto al numero de clusters. En la grafica resultante se busca el codo, es decir donde la disminucion de la suma de distancias se vuelve menos pronunciada. este punto se toma como valor optimo dado que despues mayor a eso no se agregan mejoras significativas.


10. ¿Cuál es la diferencia entre los modelos de clasificación y los modelos de regresión?

**R.** Los modelos de clasificacion asignan categorias discretas como determinar si es perro o gato. Los modelos de regresion predicen valores continuos, por ejemplo predecir el precio de una casa.

11. Menciona dos métricas de evaluación para los modelos de regresión y cómo se interpretan.

**R.** Error Cuadratico Medio de la Raiz (RMSE). Un error RMSE mas bajo indica un mejor ajuste del modelo. Las unidades de este error estan en las unidades de la variable a predecir por lo que puede ayudar en visualizar la magnitud del error.

Coeficiente de determinación (R^2). Nos dice que tan bien un modelo se ajusta a los datos. Suele ir de 0 a 1 donde 0 sera que no se ajusta mejor que la media, 1 significa que el modelo precide los datos perfectamente. Puede tener valores negativos si el modelo es muy malo.

12. ¿Qué es una matriz de confusión? ¿Cuáles métricas de evaluación podemos calcular con ella?

**R.** Es una tabla para evaluar el rendimiento de un modelo de clasificacion. Compara predicciones con valores reales.
Las metricas son: precision, recall, f1-score, support

13. Menciona las diferencias entre "correlación" y "causalidad".

**R.** Una correlacion es cuando las variables involucradas cambian en relacion la una a la otra.
Una causalidad es cuando una variable es causa de otra, es decir, tienen una conexion de causa y efecto.

14. ¿Qué es Deep Learning?

**R.** Es una rama del machine learning que usa redes neuronales artificiales con multiples capas para aprender y extraer patrones de conjuntos de datos. Es muy util en el procesamiento de voz y reconocimiento de imagenes.

15. ¿Qué es un perceptrón?


**R.** Es la unidad mas basica de una red neuronal,inspirado en el funcionamiento de las neuronas del cerebro humano. Toma entradas, las procesa mediante una funcion de activación y produce una salida. El conjunto de perceptrones forma una red neuronal.

16. ¿Qué es el procesamiento de lenguaje natural (NLP) y cómo se aplica en ciencia de datos?

**R.** Es la rama de la inteligencia artificial que se enfoca en la interaccion humano-computadora. En ciencia de datos se puede aplicar para extraer informacion de documentos de texto, reconocimiento de voz y hasta clasificacion de los anteriores para deteminar sentimientos.

17. ¿Cuál es el mejor modelo de clasificación para NLP y por qué?

**R.** El mejor modelo es el que mejor se ajuste al problema especifico, se suelen usar modelos basados en redes neuronales.

18. Menciona 3 casos de usos para Machine Learning.

**R.** Reconocimientos de imagenes, usado en el reconocimiento facial o dactilar para temas de identidad.
Sistemas de recomendacion, usado para recomendar productos o servicios que sean mas atractivos para un usuario y captar su atencion.
Diagnostico medico, para predecir o clasificar enfermedades.


19. Menciona 2 modelos de predicción y describe brevemente cómo funcionan.

**R.** Regresion lineal, ajusta los datos a una linea de modo que la suma de las distancias entre los puntos y la linea es la menor. Luego esta ecuación se usa para hacer predicciones.

 Maquina de soporte vectorial (SVM), es un modelo supervisado, se puede usar para clasificacion y para regresion el caso de la regresión el SVM encuentra la función que mejor se ajusta a los datos, minimizando la suma de los errores de predicción.



20. Menciona 2 modelos de clasificación y describe brevemente cómo funcionan.

**R.** K-Nearest Neighbors, se trata de un modelo de clasificación basado en la cercania, este agrupa los datos basandose en el espacio de caracteristicas con sus vecinos mas cercanos.

Arbol de desicion, se clasifica de acuerdo a una serie de preguntas que llevan por diferentes ramas hasta caer en la respuesta mas acertada.

### Examen Práctico


**Instrucciones**
Escoge uno de los siguientes datasets, descárgalo, define un problema a resolver y utilízalo para aplicar el modelo de Data Science que mejor creas.
* *Bike Sharing Dataset:* https://archive.ics.uci.edu/dataset/275/bike+sharing+dataset
* *Stock Market Dataset:* https://www.kaggle.com/datasets/jacksoncrow/stock-market-dataset
* *Titanic Dataset:* https://www.kaggle.com/c/titanic/data
* *Spam Email Dataset:* https://archive.ics.uci.edu/dataset/94/spambase
* *Book-Crossing Dataset:* https://www.kaggle.com/datasets/somnambwl/bookcrossing-dataset
* *Twitter Dataset:* https://www.kaggle.com/datasets/jp797498e/twitter-entity-sentiment-analysis

#### **Explicación del problema**

**R.** Solo como ejercicio interactivo, me gustaria hacer un modelo que te diga si hubieras sobrevivido o no al evento basado en tus caracteristicas como edad, clase etc. Para ello voy a desarrollar un random forest de clasificacion.

#### **Código**

In [34]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [35]:
df = pd.read_csv('train.csv')
df.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [36]:
#Empiezo por analizar que columnas me seran utiles y faciles de manejar por temas de tiempo.
unique_cabin_count = df['Cabin'].nunique()
print(f'El número de valores únicos en la columna Cabin es: {unique_cabin_count}')
unique_Embarked_count = df['Embarked'].nunique()
print(f'El número de valores únicos en la columna Embarked es: {unique_Embarked_count}')
unique_Ticket_count = df['Ticket'].nunique()
print(f'El número de valores únicos en la columna Ticket es: {unique_Ticket_count}')

El número de valores únicos en la columna Cabin es: 147
El número de valores únicos en la columna Embarked es: 3
El número de valores únicos en la columna Ticket es: 681


In [37]:
#Elimino las variables menos relevantes, o redundantes.
df.drop(columns=['Name'], inplace=True)
df.drop(columns=['Ticket'], inplace=True)
df.drop(columns=['Cabin'], inplace=True)
df.drop(columns=['PassengerId'], inplace=True)
df.head()


Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [38]:
#Cambio las variables categoricas a numericas
df['Sex'] = df['Sex'].map({'female': 0, 'male': 1})
df['Embarked'] = df['Embarked'].map({'C': 0, 'Q': 1, 'S': 2})
df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500       2.0
1         1       1    0  38.0      1      0  71.2833       0.0
2         1       3    0  26.0      0      0   7.9250       2.0
3         1       1    0  35.0      1      0  53.1000       2.0
4         0       3    1  35.0      0      0   8.0500       2.0

In [45]:
#Elimino los renglones con datos faltantes
df = df.dropna()
df.head()


Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500       2.0
1         1       1    0  38.0      1      0  71.2833       0.0
2         1       3    0  26.0      0      0   7.9250       2.0
3         1       1    0  35.0      1      0  53.1000       2.0
4         0       3    1  35.0      0      0   8.0500       2.0

In [49]:
len(df)

712

In [52]:
#Listo para hacer el modelo!
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [53]:
y = df['Survived']
X = df.drop(columns=['Survived'])

In [54]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [56]:
model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [57]:
y_pred = model.predict(X_test)

In [60]:
#Evaluamos el modelo
accuracy = accuracy_score(y_test, y_pred)

Accuracy: 0.79


In [62]:
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.84      0.82        80
           1       0.78      0.73      0.75        63

    accuracy                           0.79       143
   macro avg       0.79      0.78      0.79       143
weighted avg       0.79      0.79      0.79       143

[[67 13]
 [17 46]]


#### **Explicación de resultados**

**R.** El modelo es capaz de predecir si un usuario sobrevive o no con un 79% de efectividad. Dadas las metricas del reporte podemos decir que el modelo tiene un buen desempeño en la prediccion de supervivientes.